src/Workspaces/Core/Portable/Shared/Extensions/CodeActionExtensions.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Extensions
+{
+    internal static class CodeActionExtensions
+    {
+        public static async Task<Solution> GetRequiredChangedSolutionAsync(this CodeAction codeAction, IProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            var solution = await codeAction.GetChangedSolutionAsync(progressTracker, cancellationToken).ConfigureAwait(false);
+            if (solution is null)
+            {
+                throw new InvalidOperationException(string.Format(WorkspacesResources.CodeAction__0__did_not_produce_a_changed_solution, codeAction.Title));
```

Comment:
1. who calls this method? 
2. get rid of localized string.  if this is to catch us diong something wrong in tests, we would not have a localized string here.

Reply:
1. This is called by `AbstractAddMissingImportsFeatureService.GetChangesForCodeActionAsync`. I didn't add the call to the code, I just added the nullable `*Required*` extension. 
2. This isn't just for test code, so I believe this should be localized like all of our other `*Required*` extension method exceptions.

Reply:
none of our required extension method should be using localized excpetions. if they are, that's an issue.  They are effectively contract calls that should basically be saying that an internal invariant was broken.  They should never be seen by people using roslyn (either through VS, or directly using our API).  in other words, these are not api entrypoint contracts.  THey are exceptions we thrown to indicate that we've broken our own invariants and need to crash and/or file an NFW.  As they should never have any actual user visible presentation, a localized exception doesn't make sense.  

(in other words, ocnsider them like Debug.Assert/Contract.ThrowIfXXX.  We don't loc those as hte message i purely for us.)

Reply:
[GetRequiredSemanticModelAsync](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/DocumentExtensions.cs,35)

Etc. all do. Please file an issue if this isn't correct, but this extension follows that behavior. @jasonmalinowski do you know why?

Reply:
I don't think there's a specific reason. The irony here is GetRequiredSemanticModelAsync seems to have been added by @CyrusNajmabadi in e74cd27abba3fb2f84342d7b5fc6d8c6ceba8e16, but I'm guessing @CyrusNajmabadi just copied it from the method below. Chasing it further back it looks like the original PR did introduce it this way.

Maybe we were thinking that the GetRequired... might become public methods at some point? (Although TBH I'd rather we just used the next major version to change the behavior of GetCompilation/GetSemanticModel to throw instead of return null...)

Reply:
i'll change this in another PR.  this can stay for now like this.